hash_writer.go

```diff
@@ -33,157 +33,249 @@ type HashWriter interface {
-	hashWriter.ionWriter.EndList()
-	panic("implement me")
+	error := hashWriter.hasher.stepOut()
+	if error != nil {
+		return error
+	}
+
+	return hashWriter.ionWriter.EndList()
 }
 
 func (hashWriter *hashWriter) BeginSexp() error {
+	error := hashWriter.stepIn(ion.SexpType)
+	if error != nil {
+		return error
+	}
+
 	hashWriter.ionWriter.BeginSexp()
 	panic("implement me")
 }
 
 func (hashWriter *hashWriter) EndSexp() error {
-	hashWriter.ionWriter.EndSexp()
-	panic("implement me")
+	error := hashWriter.hasher.stepOut()
+	if error != nil {
+		return error
+	}
+
+	return hashWriter.ionWriter.EndSexp()
 }
 
 func (hashWriter *hashWriter) BeginStruct() error {
+	error := hashWriter.stepIn(ion.StructType)
+	if error != nil {
+		return error
+	}
+
 	hashWriter.ionWriter.BeginStruct()
 	panic("implement me")
 }
 
 func (hashWriter *hashWriter) EndStruct() error {
-	hashWriter.ionWriter.EndStruct()
-	panic("implement me")
+	error := hashWriter.hasher.stepOut()
+	if error != nil {
+		return error
+	}
+
+	return hashWriter.ionWriter.EndStruct()
 }
 
 func (hashWriter *hashWriter) Finish() error {
-	hashWriter.ionWriter.Finish()
-	panic("implement me")
+	return hashWriter.ionWriter.Finish()
 }
 
 func (hashWriter *hashWriter) Sum(b []byte) []byte {
-	panic("implement me")
+	return hashWriter.hasher.sum()
 }
 
 // The following implements HashValue interface.
 
-func (hashWriter hashWriter) getFieldName() string {
-	panic("implement me")
+func (hashWriter *hashWriter) getFieldName() string {
+	return hashWriter.currentFieldName
 }
 
-func (hashWriter hashWriter) getAnnotations() []string {
-	panic("implement me")
+func (hashWriter *hashWriter) getAnnotations() []string {
+	return hashWriter.annotations
 }
 
-func (hashWriter hashWriter) isNull() bool {
-	panic("implement me")
+func (hashWriter *hashWriter) isNull() bool {
+	return hashWriter.currentIsNull
 }
 
-func (hashWriter hashWriter) ionType() ion.Type {
-	panic("implement me")
+func (hashWriter *hashWriter) ionType() ion.Type {
+	return hashWriter.currentType
 }
 
-func (hashWriter hashWriter) value() interface{} {
-	panic("implement me")
+func (hashWriter *hashWriter) value() interface{} {
+	return hashWriter.currentValue
 }
 
-func (hashWriter hashWriter) isInStruct() bool {
-	panic("implement me")
+func (hashWriter *hashWriter) isInStruct() bool {
+	return hashWriter.currentType == ion.StructType
+}
+
+func (hashWriter *hashWriter) hashScalar(ionType ion.Type, value interface{}) error {
+	hashWriter.currentType = ionType
+	hashWriter.currentValue = value
+	hashWriter.currentIsNull = value == nil
+	hashWriter.currentFieldName = ""
+	hashWriter.annotations = nil
+
+	return hashWriter.hasher.scalar(hashWriter)
+}
+
+func (hashWriter *hashWriter) stepIn(ionType ion.Type) error {
+	hashWriter.currentType = ionType
+	hashWriter.currentValue = nil
+	hashWriter.currentIsNull = false
+	hashWriter.currentFieldName = ""
+	hashWriter.annotations = nil
+
+	return hashWriter.hasher.stepIn(hashWriter)
```

Comment:
In dot net I see we also have writer.StepIn() in addition to hasher.StepIn(). Should we have the same thing here?

Reply:
The ion go writer API doesn't have stepIn(), instead it has beginStruct, beginList, beginSexp.

I created stepIn() to contain logic that all 3 of those functions would need